src/Core/Content/Media/Thumbnail/ThumbnailService.php

```diff
@@ -362,25 +362,18 @@ private function calculateThumbnailSize(
 
         if ($imageSize['width'] >= $imageSize['height']) {
             $factor = $preferredThumbnailSize->getWidth() / $imageSize['width'];
-
-            $calculatedWidth = $preferredThumbnailSize->getWidth();
-            $calculatedHeight = (int) ceil($imageSize['height'] * $factor);
-
-            $useOriginalSizeInThumbnails = $imageSize['width'] < $calculatedWidth || $imageSize['height'] < $calculatedHeight;
-
-            return $useOriginalSizeInThumbnails ? [
-                'width' => $imageSize['width'],
-                'height' => $imageSize['height'],
-            ] : [
-                'width' => $calculatedWidth,
-                'height' => $calculatedHeight,
-            ];
+            if ($imageSize['height'] * $factor > $preferredThumbnailSize->getHeight()) {
```

Comment:
what do you think of extracting that logic to a separate `ThumbnailSizeCalculator`, that would make it way more easier to add unit tests for the size calculation, so you could add the testcases you described as unit tests, to make sure that we never break it again